programs/policy/policy.8
'\" t
.\"     Title: IPSEC_POLICY
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 10/06/2010
.\"    Manual: [FIXME: manual]
.\"    Source: [FIXME: source]
.\"  Language: English
.\"
.TH "IPSEC_POLICY" "8" "10/06/2010" "[FIXME: source]" "[FIXME: manual]"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ipsec_policy \- list of existing policy
.SH "SYNOPSIS"
.HP \w'\fBipsec\fR\ 'u
\fBipsec\fR \fIpolicy\fR
.SH "DESCRIPTION"
.PP
Note that policy is only supported on the new MAST stack\&. It is not supported on any other stack\&. On the
\fBklips\fR
stack, use
\fIipsec eroute\fR, on the
\fBnetkey\fR
stack, use
\fIip xfrm\fR
.PP
lists the IPSEC policy tables (aka eroutes) which control what (if any) processing is applied to non\-encrypted packets arriving for IPSEC processing and forwarding\&.
.PP
A table entry consists of:
.PP
+
.RS 4
packet count,
.RE
.PP
+
.RS 4
source address with mask and source port (0 if all ports or not applicable)
.RE
.PP
+
.RS 4
a \'\->\' separator for visual and automated parsing between src and dst
.RE
.PP
+
.RS 4
destination address with mask and destination port (0 if all ports or not applicable)
.RE
.PP
+
.RS 4
a \'=>\' separator for visual and automated parsing between selection criteria and SAID to use
.RE
.PP
+
.RS 4
SAID (Security Association IDentifier), comprised of:
.RE
.PP
+
.RS 4
protocol (\fIproto\fR),
.RE
.PP
+
.RS 4
address family (\fIaf\fR), where \'\&.\' stands for IPv4 and \':\' for IPv6
.RE
.PP
+
.RS 4
Security Parameters Index (\fISPI\fR),
.RE
.PP
+
.RS 4
effective destination (\fIedst\fR), where the packet should be forwarded after processing (normally the other security gateway) together indicate which Security Association should be used to process the packet,
.RE
.PP
+
.RS 4
a \':\' separating the SAID from the transport protocol (0 if all protocols)
.RE
.PP
+
.RS 4
source identity text string with no whitespace, in parens,
.RE
.PP
+
.RS 4
destination identity text string with no whitespace, in parens
.RE
.PP
Addresses are written as IPv4 dotted quads or IPv6 coloned hex, protocol is one of "ah", "esp", "comp" or "tun" and SPIs are prefixed hexadecimal numbers where the prefix \'\&.\' is for IPv4 and the prefix \':\' is for IPv6
.PP
SAIDs are written as "protoafSPI@edst"\&. There are also 5 "magic" SAIDs which have special meaning:
.PP
+
.RS 4
\fB%drop\fR
means that matches are to be dropped
.RE
.PP
+
.RS 4
\fB%reject\fR
means that matches are to be dropped and an ICMP returned, if possible to inform
.RE
.PP
+
.RS 4
\fB%trap\fR
means that matches are to trigger an ACQUIRE message to the Key Management daemon(s) and a hold policy will be put in place to prevent subsequent packets also triggering ACQUIRE messages\&.
.RE
.PP
+
.RS 4
\fB%hold\fR
means that matches are to stored until the policy is replaced or until that policy gets reaped
.RE
.PP
+
.RS 4
\fB%pass\fR
means that matches are to allowed to pass without IPSEC processing
.RE
.SH "EXAMPLES"
.PP
\fB1867 172\&.31\&.252\&.0/24:0 \-> 0\&.0\&.0\&.0/0:0 => tun0x130@192\&.168\&.43\&.1:0 \fR

\fB () ()\fR
.PP
means that 1,867 packets have been sent to a
\fBpolicy\fR
that has been set up to protect traffic between the subnet
172\&.31\&.252\&.0
with a subnet mask of
24
bits and the default address/mask represented by an address of
0\&.0\&.0\&.0
with a subnet mask of
0
bits using the local machine as a security gateway on this end of the tunnel and the machine
192\&.168\&.43\&.1
on the other end of the tunnel with a Security Association IDentifier of
\fBtun0x130@192\&.168\&.43\&.1\fR
which means that it is a tunnel mode connection (4, IPPROTO_IPIP) with a Security Parameters Index of
130
in hexadecimal with no identies defined for either end\&.
.PP
\fB746 192\&.168\&.2\&.110/32:0 \-> 192\&.168\&.2\&.120/32:25 => esp0x130@192\&.168\&.2\&.120:6 \fR

\fB () ()\fR
.PP
means that 746 packets have been sent to a
\fBpolicy\fR
that has been set up to protect traffic sent from any port on the host
192\&.168\&.2\&.110
to the SMTP (TCP, port 25) port on the host
192\&.168\&.2\&.120
with a Security Association IDentifier of
\fBtun0x130@192\&.168\&.2\&.120\fR
which means that it is a transport mode connection with a Security Parameters Index of
130
in hexadecimal with no identies defined for either end\&.
.PP
\fB125 3049:1::/64 \-> 0:0/0 => tun:130@3058:4::5 () ()\fR
.PP
means that 125 packets have been sent to a
\fBpolicy\fR
that has been set up to protect traffic between the subnet
\fB3049:1::\fR
with a subnet mask of
64
bits and the default address/mask represented by an address of
\fB0:0\fR
with a subnet mask of
0
bits using the local machine as a security gateway on this end of the tunnel and the machine
\fB3058:4::5\fR
on the other end of the tunnel with a Security Association IDentifier of
\fBtun:130@3058:4::5\fR
which means that it is a tunnel mode connection with a Security Parameters Index of
130
in hexadecimal with no identies defined for either end\&.
.PP
\fB42 192\&.168\&.6\&.0/24:0 \-> 192\&.168\&.7\&.0/24:0 => %passthrough\fR
.PP
means that 42 packets have been sent to a
\fBpolicy\fR
that has been set up to pass the traffic from the subnet
192\&.168\&.6\&.0
with a subnet mask of
24
bits and to subnet
192\&.168\&.7\&.0
with a subnet mask of
24
bits without any IPSEC processing with no identies defined for either end\&.
.PP
\fB2112 192\&.168\&.8\&.55/32:0 \-> 192\&.168\&.9\&.47/24:0 => %hold (east) ()\fR
.PP
means that 2112 packets have been sent to a
\fBpolicy\fR
that has been set up to hold the traffic from the host
192\&.168\&.8\&.55
and to host
192\&.168\&.9\&.47
until a key exchange from a Key Management daemon succeeds and puts in an SA or fails and puts in a pass or drop policy depending on the default configuration with the local client defined as "east" and no identy defined for the remote end\&.
.PP
\fB2001 192\&.168\&.2\&.110/32:0 \-> 192\&.168\&.2\&.120/32:0 => \fR

\fB esp0xe6de@192\&.168\&.2\&.120:0 () ()\fR
.PP
means that 2001 packets have been sent to a
\fBpolicy\fR
that has been set up to protect traffic between the host
192\&.168\&.2\&.110
and the host
192\&.168\&.2\&.120
using
192\&.168\&.2\&.110
as a security gateway on this end of the connection and the machine
192\&.168\&.2\&.120
on the other end of the connection with a Security Association IDentifier of
\fBesp0xe6de@192\&.168\&.2\&.120\fR
which means that it is a transport mode connection with a Security Parameters Index of
\fBe6de\fR
in hexadecimal using Encapsuation Security Payload protocol (50, IPPROTO_ESP) with no identies defined for either end\&.
.PP
\fB1984 3049:1::110/128 \-> 3049:1::120/128 => \fR

\fB ah:f5ed@3049:1::120 () ()\fR
.PP
means that 1984 packets have been sent to a
\fBpolicy\fR
that has been set up to authenticate traffic between the host
\fB3049:1::110\fR
and the host
\fB3049:1::120\fR
using
\fB3049:1::110\fR
as a security gateway on this end of the connection and the machine
\fB3049:1::120\fR
on the other end of the connection with a Security Association IDentifier of
\fBah:f5ed@3049:1::120\fR
which means that it is a transport mode connection with a Security Parameters Index of
\fBf5ed\fR
in hexadecimal using Authentication Header protocol (51, IPPROTO_AH) with no identies defined for either end\&.
.SH "SEE ALSO"
.PP
ipsec(8), ipsec_tncfg(5), ipsec_spi(5), ipsec_spigrp(5), ipsec_klipsdebug(5), ipsec_eroute(8), ipsec_version(5), ipsec_pf_key(5), ipsec_eroute(5)
.SH "HISTORY"
.PP
Written for the Openswan project <\m[blue]\fBhttp://www\&.openswan\&.org/\fR\m[]> by Bart Trojanowski\&.


CHANGES

Note that listings are in chronological order of release times, not in order
of version numbers, so you will find 2.5.x and 2.6.x releases intersperced.

See also docs/KNOWN_BUGS.txt and https://github.com/xelerance/Openswan/issues

v2.6.40 (February 14, 2014)
 This version specifically addresses CVE 2013-6466.

 Big changes are coming for the testing subsystem.
   From this version on, we are disengaging the testing subsystem from 
   the Openswan source tree.  You can still get a copy at
   git@github.com:xelerance/old-openswan-testing.git

 Some parts of an RFC4306/5996 patch were removed due to it
 introducing a few IKEv2 specific crashers.
 We will introduce a greater IKEv2 functionality upgrade in
 the next version.

 * CVE-2013-6466 fix: Integrated fix from Steve Lanser [Patrick Naubert]
 * KLIPS: Fix for crashes in ipsec_xmit_ipip() for 3.4.65+ kernels [Thomas Geulig]
 * Revert "relpath changes" [Brenda J. Butler]
 * Add xmlto as Debian build dependency to have fresh man pages. [Simon Deziel]
 * Avoid dns(sec) lookups for numerical sourceip= values [Paul Wouters]
 * Updated FSF address on the GPLv2 COPYING file [Paul Wouters]
 * Removed some obsoleted files in docs/ [Paul Wouters]
 * Added "ipsec initnss" command [Paul Wouters]
 * XAUTH:  Use incoming XAUTH VID when picking best connection [Andrey Alexandrenko]
 * XAUTH: fix pam race condition and contrib/pam.d file [Paul Wouters]
 * Do not perform XAUTH/ModeCfg during rekey when using Cisco compatibility [Avesh Agarwal]
 * v1phase2tov2child_integ() addition [Avesh Agarwal]
 * Changed related to bz#703985 for Secure Labeling [Avesh Agarwal]
 * Added Avesh's additional labeled ipsec logging to starterwhack [Paul Wouters]
 * Support reading NSS password from file [Paul Wouters]
 * Restore postpluto functionaliy which was missing [Tuomo Soini]
 * Don't refer to NETKEY as "2.6" or "experimental code" [Paul Wouters]
 * Added AH_SHA2_256_TRUNC to ah_transform_name_private_use [Paul Wouters]
 * helper: helper_passert_fail no longer used. Fix two string format warnings [Paul Wouters]
 * Put rpmbuild values used to compile in Makefile.inc as commented examples [Paul Wouters]
 * X509: fetch_ocsp should return void, not void * [Paul Wouters]
 * gen_reqid() can call exit_log() but confuses compiler [Paul Wouters]
 * XAUTH: fixup previous maxlength fix. mova hardcoded to defines [Paul Wouters]
 * Support /etc/sysconfig/ipsec and /etc/default/ipsec (rhbz#789917) [Paul Wouters]
 * Backporting proc_subdir_remove with Al Viro's code.
   There must a better way than me backporting something... [Patrick Naubert]
 * Added package to load dependancy for developers [Michael Richardson]
 * Make ls command explicitely avoid columns, and search both regular
   directory and execdir [Michael Richardson]
 * When logging ESP keys, be clear about which direction is which [Michael Richardson]
 * inet6 protocol does not have netns_ok flag [Michael Richardson]
 * Added netns_ok lie to get regression tests to pass [Michael Richardson]
 * Changes to work with linux 3.9 [Michael Richardson]
 * Fix a typo reported by someone to the dev@lists.openswan.org (https://lists.openswan.org/pipermail/dev/2013-September/003104.html) [Simon Deziel]
 * Update links in the README and mention that Python is a dependancy
   for ipsec verify now [Patrick Naubert]
 * Log if we send non-default PLUTO_*_RETRANSMIT_* values via env variables [Paul Wouters]
 * NETKEY: linux_pfkey_add_aead() left alg.sadb_alg_reserved uninitialised [Paul Wouters]
 * starter: remove prototypes for static functions [Paul Wouters]
 * Remove duplicate include of oswlog.h in x509dn.c [Paul Wouters]
 * Merge virtif.c header change [Paul Wouters]
 * _updown.netkey: fix route to be inserted on correct interface when
   nexthop is used [Tuomo Soini]
 * Added new option plutostderrlogtime= (default=no) [Paul Wouters]
 * Cap xauthpasslen and xauthnamelen at 128 (their buffer size) [Paul Wouters]
 * fmt_log() fix similar to previous strncat() use [Paul Wouters]
 * xauth: in theory, in xauth_inI0() it could attempt to memcpy NULL [Paul Wouters]
 * Ensure not to call same_chunk on a null pointer [Paul Wouters]
 * Simplified functions around strncat/snprintf [Paul Wouters]
 * Fixup format_end(), do not use strncat but snprintf [Paul Wouters]
 * Move the close() call for the sock to the function that created it. [Paul Wouters]
 * Undo the close on whack_sock, as it is placed in the state. [Paul Wouters]
 * Close dup()ed whack_sock in ipsecdoi_replace() to avoid leaking fd [Paul Wouters]
 * Remove other half of ipsec_copyright_notice() [Paul Wouters]
 * Include "sysdep.h" in udpfromto.c [Paul Wouters]
 * Close socket fd of the interface in _iface_down() [Paul Wouters]
 * Fix potential strncat() failure in format_end() [Paul Wouters]
 * More strnat() safety checks [Paul Wouters]
 * Additional safety checks to alg_info_snprint_esp() and
   alg_info_snprint_ah() [Paul Wouters]
 * Additional safety checks to addrtot(), inet_addrtot() and sin_addrtot() [Paul Wouters]
 * Block rules created by openswan remain even after tunnel establishment [Panagiotis Tamtamis]
 * Remove KLIPS define in initiate.c [Paul Wouters]
 * DNSSEC: added root and DLV (dlv.isc.org) key for dnssec validation [Paul Wouters]
 * ipsec-tools 0.8.0 mistakenly sets some NAT-OA fields that are defined
   in RFC1374 as "always zero". We define these as "ft_mbz" (Must Be Zero) [Paul Wouters]
 * Fixup some credits. Remove merged contrib code for selinux [Brenda J. Butler]
 * Redone and simplified functions around strncat/snprintf for addrtot.c [Paul Wouters]
 * Fix addrtot() with a passert and off-by-one [Paul Wouters]
 * Move the close() call for the sock to the function that created it. [Paul Wouters]
 * Close socket fd of the interface in _iface_down() [Paul Wouters]
 * Change name from libreswan.h to openswan.h [Brenda J. Butler]
 * Fixup IPSECKEY support with ipv4/ipv6 family and support --precedence [Paul Wouters]

programs/policy/policy.pl
#!/usr/bin/perl -w
use strict;
use warnings;
use Pod::Usage;
use Getopt::Long;
use POSIX qw( EXIT_SUCCESS EXIT_FAILURE );

# ------------------------------------------------------------------------
# This next POD section is used by pod2usage to generate --help output,
# and is used by the makefile to generate a man page.
#
=head1 NAME

ipsec_policy - show ipsec policy information

=head1 SYNOPSIS

    # detect what stack is used
    ipsec policy --detect-stack

    # display policy information
    ipsec policy [ --all | [ --inbound | --outbound | --forward ] ] \
                 [ --stack=name ] [ --read=file ] [ --debug ]

    # provide usage information
    ipsec policy --usage
    ipsec policy --help

=head1 DESCRIPTION

I<policy> displays the incoming, outgoing, and forwarding packet policies of
the system.  It is a wrapper around eixsting klips and netkey data, but
presented in a less terse form.

=head1 OPTIONS

=over 4

=item --detect-stack

Only display the stack that Libreswan is using.  Possible results are.

=over 4

=item klips

KLIPS is the Libreswan ipsec kernel module.  This stack type indicates that
KLIPS is not running in I<mast> mode (see next option), but rather in the 
default mode.  In this mode, KLIPS outgoing packet policy is dicated by
I<eroutes>.  See the B<ipsec_eroute> man page for further details.

=item mast

This is a mode of the Libreswan ipsec kernel module, KLIPS.  In this mode
outgoing packet routing policies are dictated by iptalbles, and Linux kernel
policy routing.  This mode is selected by using C<protostack=mast> setting in
I<ipsec.conf>.

=item netkey

This stack indicates that Libreswan is controlling the Linux kernel built-in
ipsec functionally.

=back


=item --all

Show inbound, outbound, and forward policites.  This is the default.

=item --inbound --in

Show only inbound policy.

=item --outbound --out

Show only outbound policy.

=item --forward --fwd

Show only forward policy.

=item --stack=<name>

Skip autodetection and force read policy from this stack.  See help on
B<--detect-stack> (above) for valid options and their descriptions.

=item --read=<file>

This option overrides what file would be read to gather the policy
information.  It could be used to read policy information from a snapshot
obtained from a running system.

In the case of the klips or mast stack, this file is the output of the
B</proc/net/ipsec/spi/all> file.

=item --help

Output help.

=item --debug

Output debug info.

=back

=head1 FILES

   /proc/net/ipsec/spi/all

=head1 SEE ALSO

B<ipsec>(8), B<ipsec_eroute>(8), B<ipsec_manual>(8)

=head1 HISTORY

Designed for the Libreswan project <http://www.libreswan.org> by Bart
Trojanowski.

=head1 BUGS

Does not support netkey yet.

=cut

# ------------------------------------------------------------------------
# general helper functions

my $debug = 0;
sub dbg
{
        my $txt = "@_";
        chomp $txt;
        warn "# $txt\n" if $debug;
}

sub assert_root
{
        my $txt = join(' ', @_);
        dbg "your uid is $>";
        if ($> ne 0) {
                die "You need to be root"
                    . (length($txt)==0 ? "." : " $txt")
                    . "\n";
        }
}

# ------------------------------------------------------------------------
# detect stack and print sadb
package Stack;
use vars qw($debug);
use subs qw(dbg assert_root);
*dbg = \&::main::dbg;
*assert_root = \&::main::assert_root;

sub detect_ipsec_stack
{
        my $class = shift;

        assert_root "to detect the ipsec stack; try --stack= and --read=.";

        my $os_mode = Stack::Klips->detect_libreswan_stack_mode();
        return $os_mode if (defined $os_mode);

        die "We don't support netkey yet.\n";
}

sub new
{
        my $class = shift;
        my (%args) = @_;

        my $stack_name = $args{name};
        $stack_name = $class->detect_ipsec_stack() if not defined $stack_name;
        dbg "stack name is $stack_name";

        my $subclass = 'Stack::' . ucfirst($stack_name);
        dbg "subclass is $subclass";

        my $instance = eval { $subclass->new(%args) };
        if($@) { die "Unknown stack: $stack_name\nERROR: $@\n" };

        $instance->initialize_spdb();

        return $instance;
}

sub initialize_spdb
{
        my $this = shift;

        $this->{spdb} = {};
        $this->{spnames} = {
                'in'  => [],
                'out' => [],
                'fwd' => [],
        };
}

sub add_spdb_entry
{
        my $this = shift;
        my ($entry) = @_;

        my $name = $entry->{name};
        ${$this->{spdb}}{$name} = $entry;

        my $dir = $entry->{dir};
        if ( $dir ne 'in' and $dir ne 'out' ) {
                die "line $. has neither dir=in nor dir=out set\n";
        }

        push @{${$this->{spnames}}{$dir}}, $name;
}

sub print_spi
{
        my $this = shift;
        my ($prefix,$s) = (@_);

        return unless ($s->{'xform'} eq 'IPIP');
        return unless (defined $s->{'policy'});

        my @p = split("->", $s->{'policy'});

        printf "%-3s  %-10d   %-18s -> %-18s     %s     ref:%d him:%d\n",
                $prefix,
                $s->{'life_packets_c'} || 0,
                $p[0],
                $p[1],
                $s->{'name'},
                $s->{'ref'},
                $s->{'refhim'};
}

sub print_spdb_dir
{
        my $this = shift;
        my ($prefix, @names) = (@_);

        for my $name (sort @names) {
                my $s = ${$this->{spdb}}{$name};
                $this->print_spi($prefix, $s);
        }
}

sub print_spdb
{
        my $this = shift;

        print STDERR "stack: " . $this->name() . "\n";

        dbg "show: " . join (',',@{$this->{show}});
        for my $what (@{$this->{show}}) {
                $this->print_spdb_dir(uc $what, @{${$this->{spnames}}{$what}});
        }
}




# ------------------------------------------------------------------------
# klips specific code
package Stack::Klips;
our @ISA = qw( Stack );
use vars qw($debug);
use subs qw(dbg assert_root);
*dbg = \&::main::dbg;
*assert_root = \&::main::assert_root;

# some well known files
my $klips_proc_version='/proc/net/ipsec/version';
my $klips_proc_spi_all='/proc/net/ipsec/spi/all';

sub have_libreswan
{
        my $class = shift;

        return 0 if system('which ipsec >/dev/null 2>&1');
        return 0 if system('ipsec version >/dev/null 2>&1');
        dbg "detected libreswan binaries";
        return 1;
}

sub have_klips_module
{
        my $class = shift;

        return 0 if not -e $klips_proc_spi_all;
        return 0 if not -e $klips_proc_version;
        return 0 if system("cat $klips_proc_version 2>/dev/null | grep -q '^Libreswan'");
        dbg "detected libreswan module";
        return 1;
}

sub detect_libreswan_stack_mode
{
        my $class = shift;

        return () if not have_libreswan;
        return () if not have_klips_module;

        return () if not open(IN, 'ipsec auto --status |');
        my $ifline = <IN>;
        close(IN);
        dbg "read: $ifline";

        return () if not $ifline =~ m/using kernel interface: (\w+)$/;
        return () if $1 ne 'klips' and $1 ne 'mast';
        dbg "detected libreswan in $1 mode";
        return $1;
}

sub new
{
        my $class = shift;
        my (%args) = @_;
        my %defaults = (
                read => $klips_proc_spi_all
        );
        my $this = { %defaults };
        bless $this, $class;
        # assign args
        while (my ($k,$v) = each(%args)) {
                next if not defined $v;
                dbg "arg $k is $v";
                $this->{$k} = $v 
        }
        return $this;
}

sub name
{
        return "klips"
}

sub read_spdb
{
        my $this = shift;

        my $input = $this->{read};
        dbg "reading from $input";
        assert_root "to read the policy database in $input." if not -r $input;

        open(SPI, "< $input") || die "failed to read $input\n";
        while (<SPI>) {
                m/^(\S+) (\S+): (.*)$/ || die "couldn't parse line $.: $_";
                my ($name, $xform, $rest) = ($1, $2, $3);

                my $tmp = { name  => $name,
                        xform => $xform };

                dbg '-' x 70;
                dbg "$name";
                dbg "$xform";
                dbg "$rest";

                foreach my $word (split /\s+/, $rest) {
                        $word =~ m/^(\S+)=(\S+)$/ || die "couldn't parse part of line $.: $word\n";
                        my ($var, $val) = ($1, $2);

                        dbg "  $var = $val";

                        $tmp->{$var} = $val;

                        if ($var eq 'life(c,s,h)') {
                                # life(c,s,h)=bytes(19240,0,0)addtime(811,0,0)usetime(784,0,0)packets(185,0,0)
                                for my $n (qw{bytes addtime usetime packets}) {
                                        if ($val =~ m/$n\((\d+),(\d+),(\d+)\)/) {
                                                dbg "    life_${n}_c = $1";
                                                dbg "    life_${n}_s = $2";
                                                dbg "    life_${n}_h = $3";

                                                $tmp->{"life_${n}_c"} = $1;
                                                $tmp->{"life_${n}_s"} = $2;
                                                $tmp->{"life_${n}_h"} = $3;
                                        }
                                }
                        }
                }

                $this->add_spdb_entry($tmp);
        }
        close(SPI);
}

# ------------------------------------------------------------------------
# mast specific code
package Stack::Mast;
our @ISA = qw( Stack::Klips );
use vars qw($debug);
use subs qw(dbg assert_root);
*dbg = \&::main::dbg;
*assert_root = \&::main::assert_root;

sub name
{
        return "mast"
}

# everything here is inhered from Klips (for now)

# ------------------------------------------------------------------------
# netkey specific code
package Stack::Netkey;
our @ISA = qw( Stack );
use vars qw($debug);
use subs qw(dbg assert_root);
*dbg = \&::main::dbg;
*assert_root = \&::main::assert_root;

sub new
{
        my $class = shift;
        my (%args) = @_;
        my %defaults = (
                # some defaults can go here
        );
        my $this = { %defaults };
        bless $this, $class;
        # assign args
        while (my ($k,$v) = each(%args)) {
                next if not defined $v;
                dbg "arg $k is $v";
                $this->{$k} = $v 
        }
        return $this;
}

sub name
{
        return "netkey"
}

sub read_spdb
{
        my $this = shift;
        print "Netkey read_spdb\n";
}

# ------------------------------------------------------------------------
# this is the main part of the program
package main;
my $show_all = 0;
my $show_inbound = 0;
my $show_outbound = 0;
my $show_forward = 0;
my $opt_stack;
my $opt_read;

my $rc = GetOptions(
        # detection only mode
        'detect-stack' => sub { print "stack=" . Stack->detect_ipsec_stack() . "\n"; exit 0 },
        # what to show
        'all'       => \$show_all,
        'inbound'   => \$show_inbound,
        'in'        => \$show_inbound,
        'outbound'  => \$show_outbound,
        'out'       => \$show_outbound,
        'forward'   => \$show_forward,
        'fwd'       => \$show_forward,
        # what to read
        'stack=s'   => \$opt_stack,
        'read=s'    => \$opt_read,
        # generic stuff
        'debug'     => \$debug,
        # help
        'usage'     => sub { pod2usage(-exitval => EXIT_SUCCESS, -verbose => 0) },
        'help'      => sub { pod2usage(-exitval => EXIT_SUCCESS, -verbose => 1) },
);
pod2usage('Invalid options specified.') if !$rc;

if ($show_all) {
        # show everything
        $show_inbound = $show_outbound = $show_forward = 1;

} elsif (not ($show_inbound | $show_outbound | $show_forward)) {
        # by default show only outbound policies
        $show_outbound = 1;
}

# collect names of policies to display
my @show = ();
push @show, 'in'  if $show_inbound;
push @show, 'out' if $show_outbound;
push @show, 'fwd' if $show_forward;

# get a new stack object
my $s = Stack->new(
        name => $opt_stack,
        read => $opt_read,
        show => \@show,
);

$s->read_spdb();
$s->print_spdb();

